### ISSUE TITLE
NuProj.targets Clean and Build targets do not support CleanDependsOn and BuildDependsOn property convention for extension

### ISSUE THREAD

**Issue by NicoleTedesco** [terrajobst/nuproj#60]
The standard Microsoft Build, Clean, and Rebuild targets use a "xxxDependsOn" property convention in order to allow local projects to properly extend the original target's behavior.  The NuProj target overrides do not follow this convention, which forces the developer to completely copy the original NuProj Clean and Build targets into the local build in order to extend them.

http://msdn.microsoft.com/en-us/library/ms366724.aspx

**Comment by kovalikp** [comment]
Hi Nicole. You can use `BeforeTargets` and `AfterTargets` attributes to extend the build script.

**Comment by NicoleTedesco** [comment]
Not if, as NuProj.targets have done, the Clean, Build and Rebuild targets have been completely overwritten.  BeforeBuild and AfterBuild work because they are included as dependencies in Microsoft's definition of the original targets.

**Comment by williamb1024** [comment]
@NicoleTedesco. kovalikp is suggesting that you use the `AfterTargets` and `BeforeTargets` attributes of the `Target` element. 

[Target Element (MSBuild)](http://msdn.microsoft.com/en-us/library/t50z2hka.aspx)

These attributes will let you sequence your target with a different target without modifying the predefined target.

**Comment by NicoleTedesco** [comment]
Not true.  That did not work, which motivated me to search the conventions of build extension.  The problem is that BeforeBuild works and AfterBuild works precisely because Microsoft built them in to their original target definition for "Build".  The problem with the NuProj build is that the tasks for "Build", "Clean" and "Rebuild" have been COMPLETELY REDEFINED!  The entire build extensions system that Microsoft built has been cut short.  I wrote the approved changes into my "patch-1" version of the file, which is similar to what Microsoft did, and what must be done here as well.  Otherwise, neither "AfterClean" nor "AfterBuild" work at all.

Nothing.  Nada.  Zip.

**Comment by NicoleTedesco** [comment]
I think the only difference between Microsoft's built-in system and my patch is that BeforeXxx and AfterXxx are invoked only if they exist.  I added dummy targets anyway, which can always be overridden to perform something meaningful.

**Comment by NicoleTedesco** [comment]
I wasn't sure how to turn this into tests.  I really didn't have time at work to devise a custom build sequence.  However, I did solve this problem in a local build at work.  I performed a copy-and-paste into my branched file.

**Comment by NicoleTedesco** [comment]
(BTW, I needed to implement both custom post-clean tasks and post-build tasks for automated deployment.)

**Comment by NicoleTedesco** [comment]
Neither BeforeXxx or AfterXxx are magical targets.  They need to be setup to be useful.

**Comment by NicoleTedesco** [comment]
BTW, I have some property functions that build properties $(NumberOfDaysSinceEpoch) and $(NumberOfSecondsSinceMidnight) if anyone is interested.  I used them to create automatic build and revision numbers for NuGet packages.  (This is similar to the major.minor.* convention for automated assembly versioning.)

**Comment by williamb1024** [comment]
I understand what you are saying. You can achieve the same result by adding:

`<Target Name="MyAfterClean" AfterTargets="Clean">
    
</Target>`

Creating your target this way, you can do so without having to rely on changes to the `NuProj.targets` file.

The `CleanDependsOn` properties and such, I believe are relics from pre-`AfterTargets` and `BeforeTargets` support (which was introduced in MSBuild 4.0)

**Comment by NicoleTedesco** [comment]
I will test them out tomorrow.  Thanks.

**Comment by NicoleTedesco** [comment]
Ah, well, I'll be darned!  Learn something new every day.  I will DEFINITELY be sure to test this tomorrow.

Thanks,

https://freetodev.wordpress.com/2009/06/24/msbuild-4-0-beforetargets-and-aftertargets/

**Comment by NicoleTedesco** [comment]
I wonder if BOTH should be supported, since Microsoft supports both and had not hinted that the xxxDependsOn conventions will be deprecated any time soon.

**Comment by Michael-DST** [comment]
FWIW, thank you for this thread.  I was just like Nicole and under the impression there needed to be an after/before defined in the primary .proj file to override.  It might be useful to add blank/empty hooks in the created NuProj project file as a convenience.

**Comment by NN---** [comment]
+1
I would like to see normal BeforeBuild and AfterBuild.

You can meanwhile add to the template this:

```xml
<!-- To modify your build process, add your task inside one of the targets below and uncomment it. 
<Target Name="BeforeBuild" BeforeTargets="Build">
</Target>
<Target Name="AfterBuild" AfterTargets="Build">
</Target>
-->
```

**Comment by NN---** [comment]
I have added PR fixing this issue.
Can you merge it ?
Thanks.
